src/Sulu/Bundle/AdminBundle/Resources/js/components/Tabs/Tabs.js

```diff
@@ -1,56 +1,398 @@
 // @flow
 import React from 'react';
-import type {ChildrenArray, Element} from 'react';
+import type {ChildrenArray, Element, ElementRef} from 'react';
+import {action, computed, observable} from 'mobx';
+import {observer} from 'mobx-react';
+import classNames from 'classnames';
+import debounce from 'debounce';
+import Popover from '../Popover';
+import Icon from '../Icon';
+import type {Skin} from './types';
 import Tab from './Tab';
+import CollapsedTabList from './CollapsedTabList';
+import CollapsedTab from './CollapsedTab';
 import tabsStyles from './tabs.scss';
 
 type Props = {
     children: ChildrenArray<Element<typeof Tab> | false>,
     onSelect: (tabIndex: number) => void,
     selectedIndex: ?number,
+    skin: Skin,
+    small: boolean,
 };
 
-export default class Tabs extends React.PureComponent<Props> {
+const DEBOUNCE_TIME = 200;
+
+@observer
+class Tabs extends React.Component<Props> {
+    static defaultProps = {
+        skin: 'default',
+        small: false,
+    };
+
+    @observable wrapperWidth: number = 0;
+    @observable tabsWrapperWidth: number = 0;
+    @observable tabsWidth: number = 0;
+
+    @observable tabWidths: Map<number, number> = new Map();
+    @observable tabRefs: Map<number, ?ElementRef<'li'>> = new Map();
+    @observable dropdownOpen = false;
+    @observable lastSelectedIndex: ?number;
+
     static Tab = Tab;
 
+    resizeObserver: ?ResizeObserver;
+
+    containerRef: ?ElementRef<'div'>;
+    wrapperRef: ?ElementRef<'div'>;
+    tabsWrapperRef: ?ElementRef<'div'>;
```

Comment:
okay:
- the container is needed as anchor for the dropdown.
- the wrapper is needed for calculation of the shown tabs. so if the width of all tabs is smaller than the wrapperWidth, all tabs can be shown and the dropdownHandle is hidden. because the container has a horizontal padding, i can't use that one instead.
- the tabsWrapper is needed for calculation of the shown tabs, if there are collapsed tabs, because it's width equals to wrapperWidth minus dropdownHandleWidth

I hope that clarifies that a bit. I know that the naming is really bad, do you have any suggestions?

Reply:
> * the container is needed as anchor for the dropdown.

Still not sure if we shouldn't use the button revealing the dropdown as anchor, since that is where the `Popover` should really be attached to. This way me might get rid of that container and not require the high `horizontalOffset` prop value.

> * the wrapper is needed for calculation of the shown tabs. so if the width of all tabs is smaller than the wrapperWidth, all tabs can be shown and the dropdownHandle is hidden. because the container has a horizontal padding, i can't use that one instead.
> * the tabsWrapper is needed for calculation of the shown tabs, if there are collapsed tabs, because it's width equals to wrapperWidth minus dropdownHandleWidth
> 
> I hope that clarifies that a bit. I know that the naming is really bad, do you have any suggestions?

But let's assume we need all of these elements (and the corresponding CSS classes). I think I would start by calling the most outer `div` just `tabsContainer`, since that is the DOM element that represents the entire `Tabs` component. Since the most inner one is the one just wrapping some `Tab` components, I would call that one `tabContainer`. The one in between could then be called just `tabs`. I think that would at least make it a little bit clearer, although it is still not perfect :see_no_evil: If you want I can try a bit around, and see if I find something fitting, including if we could get rid of one container.